### ISSUE TITLE
Add proper documentation in the crate

### ISSUE THREAD

**Issue by bbqsrc** [cucumber-rs/cucumber#63]
Having a useful readme is nice, having useful crate documentation is even better.

**Comment by bbros-dev** [comment]
Note sure if this is the right issue to track this....

It would be useful to know the cucmber-rust counterparts to the Cucumber documentation.

Example: What is cucumber-rust statement that addresses "how-cucumber-finds-your-features-and-step-definitions"

https://cucumber.io/docs/gherkin/step-organization/#how-cucumber-finds-your-features-and-step-definitions

The upstream docs might also predefine your documentation structure, as well as topics and the detail-scope of the documentation. 

Upstream have worked long and hard on this - seems sensible to build on their experience and avoid relearning their mistakes?

**Comment by bbqsrc** [comment]
I appreciate your contribution, and I'd say right now, this is the right place for this type of discussion. :smile:

I think what you suggest is quite reasonable! I am quite bad at documentation planning, so if you want to help, we could start with someone perhaps yourself listing some bullet points on the kinds of topics we should cover relative to the Cucumber documentation.

**Comment by bbros-dev** [comment]
Okay, still a rust noob, but doc might be where I can contribute.

How does the following sound:

- there are two categories of documentation:
  1. Those found on [docs.rs](https://docs.rs]: Written within `*.rs` files, generated by doc.rs/rustdoc/cargo doc, aimed at **contributors to** `cucumber_rust` crate and assumes high level of Rust and possibly cucumber/gherkin knowledge.  Limited links to upstream, examples are limits to the function being documented and are not complete working exmaples of cucumber usage.
  2. Those found on GitHub-Pages (or some such): Written within `*.md` files, generated by githuib or a like static site generator tool, aimed at **users of** cucumber-rust and assumes no minimum level of Rust not Cucumber/Gherkin.  Links to upstream Cucumber/Gherkin documentation as much as possible. Callsout where/how cucmber-rust reviates from upstream behaviour. Extensive use of complete end-user examples.

Thoughts?

**Comment by bbqsrc** [comment]
My thoughts are:

We have a book in the works! You can definitely help us with that, that's the item 2 there. Once there's a skeleton, I'd be happy for you to offer as many of those things as you want. :smile:

**Comment by bbros-dev** [comment]
@bbqsrc, upstream sound receptive to having user docs (type 2, per [this comment](https://github.com/bbqsrc/cucumber-rust/issues/63#issuecomment-789489086)) land there.  

Still working out some of the details. Eventually there would be a "Rust" icon you'd click on to see the user related documentation that is specific for this crate, e.g. https://cucumber.io/docs/cucumber/api/

Since this issue is about crate documentation, I suggest opening a new issue and tracking user documentation progress there.

> We have a book in the works! ... Once there's a skeleton, I'd be happy for you to offer as many of those things as you want.

Happy to help on the book - I assume the book will bee some CC license?  Personally I don't see a conflict, esp if the CC license is compatible with the website.  The upstream docs are quite terse in terms of language specifics - more of a reference source. The book would be free to be a deeper dive into usage in a Rust context.

@bbqsrc do you agree to such an approach?

**Comment by bbqsrc** [comment]
Oh I hadn't even thought about our content ending up on cucumber.io yet. We're not 1.0, and I won't commit to an API so soon. But we will have a published github pages mdbook. (Working on it right now!)

The book would be under the same license as the codebase right now, which is MIT or Apache-2.0. But I think @alvskar wouldn't be against relicensing to CC BY or whatever the most suitable documentation license is.

**Comment by bbros-dev** [comment]
> Oh I hadn't even thought about our content ending up on cucumber.io yet. We're not 1.0, and I won't commit to an API so soon.

Okay lets wait to 1.0

> The book would be under the same license as the codebase right now

Great. Just checked and upstream docs are MIT so no need to change that.

**Comment by bbqsrc** [comment]
Okay, the early alpha book is published here now: https://bbqsrc.github.io/cucumber-rust/

The .md files are in the `book/` directory. Feel free to go wild. :smile:

**Comment by bbqsrc** [comment]
@bbros-dev tracking issue for user documentation: https://github.com/bbqsrc/cucumber-rust/issues/110

Let's continue there.

**Comment by tyranron** [comment]
Mostly resolved in #128, some biskeshedding still needs to be done before a 0.10 release, though.
